### PR TITLE
Improve error handling and tests of lockapi/mmapi

### PR DESCRIPTION
**Pull request by Martchus**
See https://progress.opensuse.org/issues/80264 and particular commit messages

### CODE REVIEW

File: t/30-mmapi.t [hunk 1]
```diff
@@ -55,35 +102,104 @@ $fake_api->get('/workers' => sub {
...
+    is(lockapi::mutex_unlock('unlockable'),          1,                                           'mutex unlocked');
+    is(lockapi::barrier_create('lucky_barrier', 41), 1,                                           'barrier created');
+    is(lockapi::barrier_wait('unblocked'),           1,                                           'waited for barrier');
+    is($recorded_info[0]->[1],                       'Wait for unblocked (on parent job)',        'info recorded by waited for barrier');
+    is(lockapi::barrier_wait({name => 'check_dead_job_barrier', check_dead_job => 1}), 1,         'waited for barrier with check_dead_job flag');
+    is($recorded_info[1]->[1],                 'Wait for check_dead_job_barrier (on parent job)', 'different info recorded with check_dead_job flag');
+    is(lockapi::barrier_try_wait('unblocked'), 1,                                                 'tried waiting for barrier');
+    is(lockapi::barrier_destroy('deletable'),  1,                                                 'barrier destroyed');
+    is(scalar @recorded_info,                  2, 'record info called expected number of times') or diag explain \@recorded_info;
```
Comment (okurz) [inline t/30-mmapi.t:202]: the indentation looks a bit unusual. If you get rid of the parentheses for Test::More calls this IMHO looks better

```suggestion
    is lockapi::mutex_lock('lockable'),     1, 'mutex locked';
    is lockapi::mutex_try_lock('lockable'), 1, 'mutex locked (try)';
    is lockapi::mutex_unlock('unlockable'), 1, 'mutex unlocked';
    is lockapi::barrier_create('lucky_barrier', 41), 1, 'barrier created';
    is lockapi::barrier_wait('unblocked'), 1, 'waited for barrier';
    is $recorded_info[0]->[1], 'Wait for unblocked (on parent job)', 'info recorded by waited for barrier';
    is lockapi::barrier_wait({name => 'check_dead_job_barrier', check_dead_job => 1}), 1, 'waited for barrier with check_dead_job flag';
    is $recorded_info[1]->[1], 'Wait for check_dead_job_barrier (on parent job)', 'different info recorded with check_dead_job flag';
    is lockapi::barrier_try_wait('unblocked'), 1, 'tried waiting for barrier';
    is lockapi::barrier_destroy('deletable'),  1, 'barrier destroyed';
    is scalar @recorded_info, 2, 'record info called expected number of times' or diag explain \@recorded_info;
```

File: lockapi.pm [hunk 2]
```diff
@@ -25,7 +25,7 @@ our @EXPORT = qw(mutex_create mutex_lock mutex_unlock mutex_try_lock mutex_wait
 
 require bmwqemu;
 use mmapi qw(api_call get_job_info);
-use testapi 'record_info';
+use testapi;
```
Comment (okurz) [inline lockapi.pm:28]: I think an explicit `use testapi ()` can show the intention better

File: t/30-mmapi.t [hunk 3]
```diff
@@ -20,11 +20,50 @@
...
+
 use OpenQA::Test::TimeLimit '5';
 use Test::Output;
 use Test::MockModule;
 use Mojolicious;
 use mmapi;
+use lockapi;
+use testapi;
+use autotest;
+use basetest;
```
Comment (okurz) [inline t/30-mmapi.t:36]: You don't need any of these imports if you just mock them. I tried it out, deleted all three, tests still pass :)

**Review by okurz — CHANGES_REQUESTED** [review]
1. can you explain why you remove t/09-lockapi.t? I see it mentioned in the git commit message of https://github.com/os-autoinst/os-autoinst/pull/1578/commits/a51148739d3365b949744b646b4114f5901a8dee but it does not state a reason. I think it would be beneficial to still have a very limited unit test for lockapi only.
2. I see uncovered test output:

```
$ prove -I. -Iexternal/os-autoinst-common/lib t/30-mmapi.t 
t/30-mmapi.t .. 1/? [2020-12-01T16:37:40.923 CET] [debug] mutex create 'lucky_lock'
[2020-12-01T16:37:40.926 CET] [debug] mutex lock 'lockable'
[2020-12-01T16:37:40.929 CET] [debug] mutex try lock 'lockable'
[2020-12-01T16:37:40.932 CET] [debug] mutex unlock 'unlockable'
[2020-12-01T16:37:40.935 CET] [debug] barrier create 'lucky_barrier' for 41 tasks
[2020-12-01T16:37:40.938 CET] [debug] barrier wait 'unblocked'
[2020-12-01T16:37:40.942 CET] [debug] barrier wait 'check_dead_job_barrier'
[2020-12-01T16:37:40.945 CET] [debug] barrier try wait 'unblocked'
[2020-12-01T16:37:40.948 CET] [debug] barrier destroy 'deletable'
t/30-mmapi.t .. ok   
All tests successful.
Files=1, Tests=6,  0 wallclock secs ( 0.04 usr  0.00 sys +  0.58 cusr  0.03 csys =  0.65 CPU)
Result: PASS
```

File: mmapi.pm [hunk 4]
```diff
@@ -106,16 +108,25 @@ Queries openQA's multi-machine API and returns the result as Mojo::Message::Resp
 
 sub api_call { api_call_2(@_)->res }
 
-sub _handle_api_error {
-    my ($tx, $log_ctx) = @_;
+=head2 handle_api_error
+
+Returns a truthy value if the specified Mojo::Transaction::HTTP object has an error.
+Logs the errors if a log context specified.
```
Comment (kalikiana) [inline mmapi.pm:114]:
```suggestion
Logs the errors if a log context was specified.
```

**Review by kalikiana — APPROVED** [review]
I like the clean-up, especially also the new variables that can be overridden. Might turn out handy outside of tests as well.

**Comment by Martchus** [comment]
> can you explain why you remove t/09-lockapi.t?

This test was just bad, e.g. it pretends to check whether parameters are passes as expected but actually doesn't. It also generally doesn't check the logging. It was also written in a very verbose way. So I decided to rewrite it and include it into the mmapi test.

> I see uncovered test output

Ok, I'll try to fix that and the other things you've mentioned.

File: lockapi.pm [hunk 5]
```diff
@@ -24,23 +24,30 @@ our @EXPORT = qw(mutex_create mutex_lock mutex_unlock mutex_try_lock mutex_wait
...
+    my $log_ctx               = "acquiring $type '$name'";
+    my %expected_return_codes = (200 => 1, 409 => 1, 410 => 1);
+    my $actual_return_code;
+    for (1 .. RETRY_COUNT) {
+        my $tx = api_call_2(post => "$type/$name", $param, \%expected_return_codes);
+        $actual_return_code = $tx->res->code;
+        last unless mmapi::handle_api_error($tx, $log_ctx, \%expected_return_codes);
+        last unless $actual_return_code == 410;
+        bmwqemu::fctinfo("Retry $_ of " . RETRY_COUNT);
+        sleep RETRY_INTERVAL;
```
Comment (okurz) [inline lockapi.pm:46]: codecov reports this as uncovered. Please try to cover that or mark with `# uncoverable statement`

File: lockapi.pm [hunk 6]
```diff
@@ -169,20 +171,16 @@ sub barrier_wait {
...
 sub barrier_destroy {
     my ($name, $where) = @_;
     bmwqemu::mydie('missing barrier name') unless $name;
-    bmwqemu::diag("barrier destroy '$name'");
-    my $param;
-    $param->{where} = $where if $where;
-    my $res = api_call('delete', "barrier/$name", $param)->code;
-    return 1 if ($res == 200);
-    bmwqemu::fctwarn("Unknown return code $res for lock api");
+    return _api_call_with_logging_and_error_hanndling("barrier destroy '$name'",
```
Comment (okurz) [inline lockapi.pm:182]: s/hanndling/handling/

**Review by okurz — CHANGES_REQUESTED** [review]

File: lockapi.pm [hunk 7]
```diff
@@ -24,23 +24,30 @@ our @EXPORT = qw(mutex_create mutex_lock mutex_unlock mutex_try_lock mutex_wait
...
+    my $log_ctx               = "acquiring $type '$name'";
+    my %expected_return_codes = (200 => 1, 409 => 1, 410 => 1);
+    my $actual_return_code;
+    for (1 .. RETRY_COUNT) {
+        my $tx = api_call_2(post => "$type/$name", $param, \%expected_return_codes);
+        $actual_return_code = $tx->res->code;
+        last unless mmapi::handle_api_error($tx, $log_ctx, \%expected_return_codes);
+        last unless $actual_return_code == 410;
+        bmwqemu::fctinfo("Retry $_ of " . RETRY_COUNT);
+        sleep RETRY_INTERVAL;    # uncoverable statement
```
Comment (okurz) [inline lockapi.pm:46]: thing is: If this line is uncovered at least the one above is as well. I haven't checked the complete coverage report, just tried to explain the CI coverage analysis comment. Can you please crosscheck for complete coverage if feasible or mark uncoverable statements accordingly?

**Review by okurz — CHANGES_REQUESTED** [review]

**Comment by Martchus** [comment]
I've marked all lines which are hard to cover (and weren't covered by the previous test as well) as uncoverable. The only exception are the functions `wait_for_children` and `wait_for_children_to_start`. They were covered at all (also not by the previous test) so I thought it is best to actually cover them.

**Review by okurz — APPROVED** [review]

**Comment by okurz** [comment]
Awesome work here. I have verified enough locally that the new test additions work fine so I can merge.

**Pull request merged**